homeassistant/components/toon/coordinator.py

```diff
@@ -0,0 +1,132 @@
+
+from .const import CONF_CLOUDHOOK_URL, DEFAULT_SCAN_INTERVAL, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ToonDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching WLED data from single endpoint."""
+
+    def __init__(
+        self, hass: HomeAssistant, *, entry: ConfigEntry, session: OAuth2Session
+    ):
+        """Initialize global Toon data updater."""
+        self.session = session
+        self.entry = entry
+
+        async def async_token_refresh() -> str:
+            await session.async_ensure_token_valid()
+            return session.token["access_token"]
+
+        self.toon = Toon(
+            token=session.token["access_token"],
+            session=async_get_clientsession(hass),
+            token_refresh_method=async_token_refresh,
+        )
+
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=DEFAULT_SCAN_INTERVAL
+        )
+
+    def update_listeners(self) -> None:
+        """Call update on all listeners."""
+        for update_callback in self._listeners:
+            update_callback()
+
+    async def register_webhook(self, event: Optional[Event] = None) -> None:
+        """Register a webhook with Toon to get live updates."""
+        if CONF_WEBHOOK_ID not in self.entry.data:
+            data = {**self.entry.data, CONF_WEBHOOK_ID: secrets.token_hex()}
+            self.hass.config_entries.async_update_entry(self.entry, data=data)
+
+        if self.hass.components.cloud.async_active_subscription():
+
+            if CONF_CLOUDHOOK_URL not in self.entry.data:
+                webhook_url = await self.hass.components.cloud.async_create_cloudhook(
+                    self.entry.data[CONF_WEBHOOK_ID]
+                )
+                data = {**self.entry.data, CONF_CLOUDHOOK_URL: webhook_url}
+                self.hass.config_entries.async_update_entry(self.entry, data=data)
+            else:
+                webhook_url = self.entry.data[CONF_CLOUDHOOK_URL]
+        else:
+            webhook_url = self.hass.components.webhook.async_generate_url(
+                self.entry.data[CONF_WEBHOOK_ID]
+            )
+
+        webhook_register(
+            self.hass,
+            DOMAIN,
+            "Toon",
+            self.entry.data[CONF_WEBHOOK_ID],
+            self.handle_webhook,
+        )
+
+        try:
+            await self.toon.subscribe_webhook(
+                application_id=self.entry.entry_id, url=webhook_url
+            )
+            _LOGGER.info("Registered Toon webhook: %s", webhook_url)
+        except ToonError as err:
+            _LOGGER.error("Error during webhook registration - %s", err)
+
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, self.unregister_webhook
+        )
+
+    async def handle_webhook(
+        self, hass: HomeAssistant, webhook_id: str, request
+    ) -> None:
+        """Handle webhook callback."""
```

Comment:
Is there any validation we can add to make sure the caller is the expected one?

Reply:
Uuh... Maybe, will check, good point.

Reply:
There is no match to that. The webhook IDs are considered unique and only can change data (which will be updated again automatically). If t a request doesn't provide any matching data, it will simply be ignored.

Reply:
I've added additional validation to the data received. Unfortunately, it does not provide actual checksums or the signing of data. At least I've added checks to see if the required fields are there and the data received is from the Toon display we are tracking (by name).